File: src/panoptikon/api/routers/jobs.py
import logging
import multiprocessing
import threading
from dataclasses import dataclass
from typing import Any, Dict, List, Literal, Optional

from fastapi import APIRouter, Body, Depends, HTTPException, Query, status
from pydantic import BaseModel

from panoptikon.api.routers.jobs_impl import (
    delete_model_data,
    rescan_folders,
    run_data_extraction_job,
    run_folder_update,
)
from panoptikon.api.routers.utils import get_db_readonly, get_db_system_wl
from panoptikon.data_extractors.models import ModelOptsFactory
from panoptikon.db import get_database_connection
from panoptikon.db.config import persist_system_config, retrieve_system_config
from panoptikon.db.extraction_log import get_all_data_logs
from panoptikon.db.files import get_all_file_scans
from panoptikon.db.folders import get_folders_from_database
from panoptikon.types import FileScanRecord, LogRecord, SystemConfig

logger = logging.getLogger(__name__)

router = APIRouter(
    prefix="/api/jobs",
    tags=["jobs"],
    responses={404: {"description": "Not found"}},
)

# Define Job Types
JobType = Literal[
    "data_extraction",
    "data_deletion",
    "folder_rescan",
    "folder_update",
]


class Job(BaseModel):
    queue_id: int
    job_type: JobType
    conn_args: Dict[str, Any]
    metadata: Optional[str] = None
    included_folders: Optional[List[str]] = None
    excluded_folders: Optional[List[str]] = None


@dataclass
class RunningJob:
    job: Job
    process: multiprocessing.Process


class QueueStatusModel(BaseModel):
    running_job: Optional["RunningJobModel"]
    queue: List["JobModel"]


class RunningJobModel(BaseModel):
    queue_id: int
    job_type: JobType
    metadata: Optional[str] = None


class JobModel(BaseModel):
    queue_id: int
    job_type: JobType
    metadata: Optional[str] = None


def execute_job(job: Job):
    try:
        if job.job_type == "data_extraction":
            assert job.metadata is not None, "Inference ID is required."
            model = ModelOptsFactory.get_model(job.metadata)
            run_data_extraction_job(model=model, conn_args=job.conn_args)
        elif job.job_type == "data_deletion":
            assert job.metadata is not None, "Inference ID is required."
            model = ModelOptsFactory.get_model(job.metadata)
            delete_model_data(model=model, conn_args=job.conn_args)
        elif job.job_type == "folder_rescan":
            rescan_folders(conn_args=job.conn_args)
        elif job.job_type == "folder_update":
            assert (
                job.included_folders is not None
                and job.excluded_folders is not None
            ), "Both included and excluded folders are required."
            run_folder_update(
                included_folders=job.included_folders,
                excluded_folders=job.excluded_folders,
                conn_args=job.conn_args,
            )
        else:
            logger.error(f"Unknown job type: {job.job_type}")
    except Exception as e:
        logger.error(f"Job {job.queue_id} failed with error: {e}")


class JobManager:
    def __init__(self):
        self.job_queue: List[Job] = []
        self.running_job: Optional[RunningJob] = None
        self.queued_jobs: Dict[int, Job] = {}
        self.job_counter: int = 0
        self.lock = threading.Lock()
        self.worker_thread = threading.Thread(
            target=self.job_consumer, daemon=True
        )
        self.worker_thread.start()
        logger.info("JobManager initialized and worker thread started.")

    def get_next_job_id(self) -> int:
        with self.lock:
            self.job_counter += 1
            return self.job_counter

    def enqueue_job(self, job: Job):
        with self.lock:
            self.job_queue.append(job)
            self.queued_jobs[job.queue_id] = job
            logger.info(f"Enqueued job {job.queue_id}: {job.job_type}")

    def job_consumer(self):
        while True:
            job = None
            with self.lock:
                if self.running_job is None and self.job_queue:
                    job = self.job_queue.pop(0)
                    self.queued_jobs.pop(job.queue_id, None)
                    logger.info(f"Dequeued job {job.queue_id}: {job.job_type}")

            if job:
                process = multiprocessing.Process(
                    target=execute_job, args=(job,)
                )
                running_job = RunningJob(job=job, process=process)
                with self.lock:
                    self.running_job = running_job
                logger.info(
                    f"Starting job {job.queue_id} in process {process.pid}"
                )
                process.start()
                process.join()
                with self.lock:
                    if (
                        self.running_job
                        and self.running_job.job.queue_id == job.queue_id
                    ):
                        logger.info(f"Job {job.queue_id} completed.")
                        self.running_job = None
            else:
                # No job to process, sleep briefly to prevent tight loop
                threading.Event().wait(1)

    def get_queue_status(self) -> QueueStatusModel:
        with self.lock:
            queue_list = [
                JobModel(
                    queue_id=job.queue_id,
                    job_type=job.job_type,
                    metadata=job.metadata,
                )
                for job in self.job_queue
            ]
            running = (
                RunningJobModel(
                    queue_id=self.running_job.job.queue_id,
                    job_type=self.running_job.job.job_type,
                    metadata=self.running_job.job.metadata,
                )
                if self.running_job
                else None
            )
        return QueueStatusModel(running_job=running, queue=queue_list)

    def cancel_queued_jobs(self, queue_ids: List[int]) -> List[int]:
        cancelled = []
        with self.lock:
            for qid in queue_ids:
                job = self.queued_jobs.pop(qid, None)
                if job and job in self.job_queue:
                    self.job_queue.remove(job)
                    cancelled.append(qid)
                    logger.info(f"Cancelled queued job {qid}: {job.job_type}")
        return cancelled

    def cancel_running_job(self) -> Optional[int]:
        with self.lock:
            if self.running_job:
                pid = self.running_job.process.pid
                self.running_job.process.terminate()
                self.running_job.process.join()
                logger.info(
                    f"Cancelled running job {self.running_job.job.queue_id} with PID {pid}"
                )
                completed_job_id = self.running_job.job.queue_id
                self.running_job = None
                return completed_job_id
            else:
                return None


# Initialize FastAPI app and JobManager
job_manager = JobManager()


# Endpoint to get queue status
@router.get(
    "/queue",
    summary="Get running job and queue status",
)
def get_queue_status() -> QueueStatusModel:
    return job_manager.get_queue_status()


# Endpoint to run a data extraction job
@router.post(
    "/data/extraction",
    summary="Run a data extraction job",
)
def enqueue_data_extraction_job(
    inference_ids: List[str] = Query(..., title="Inference ID List"),
    conn_args: Dict[str, Any] = Depends(get_db_system_wl),
) -> List[JobModel]:
    jobs = []
    for inference_id in inference_ids:
        queue_id = job_manager.get_next_job_id()
        job = Job(
            queue_id=queue_id,
            job_type="data_extraction",
            conn_args=conn_args,
            metadata=inference_id,
        )
        job_manager.enqueue_job(job)
        jobs.append(
            JobModel(
                queue_id=job.queue_id,
                job_type=job.job_type,
                metadata=job.metadata,
            )
        )
    return jobs


# Endpoint to delete extracted data
@router.delete(
    "/data/extraction",
    summary="Delete extracted data",
    status_code=status.HTTP_202_ACCEPTED,
)
def enqueue_delete_extracted_data(
    inference_ids: str = Query(..., title="Inference ID List"),
    conn_args: Dict[str, Any] = Depends(get_db_system_wl),
) -> List[JobModel]:
    jobs = []
    for inference_id in inference_ids:
        queue_id = job_manager.get_next_job_id()
        job = Job(
            queue_id=queue_id,
            job_type="data_deletion",
            conn_args=conn_args,
            metadata=inference_id,
        )
        job_manager.enqueue_job(job)
        jobs.append(
            JobModel(
                queue_id=job.queue_id,
                job_type=job.job_type,
                metadata=job.metadata,
            )
        )
    return jobs


# Endpoint to run a folder rescan
@router.post(
    "/folders/rescan",
    summary="Run a folder rescan",
    status_code=status.HTTP_202_ACCEPTED,
)
def enqueue_folder_rescan(
    conn_args: Dict[str, Any] = Depends(get_db_system_wl),
) -> JobModel:
    queue_id = job_manager.get_next_job_id()
    job = Job(
        queue_id=queue_id,
        job_type="folder_rescan",
        conn_args=conn_args,
    )
    job_manager.enqueue_job(job)
    return JobModel(
        queue_id=job.queue_id, job_type=job.job_type, metadata=job.metadata
    )


class Folders(BaseModel):
    included_folders: List[str]
    excluded_folders: List[str]


# Endpoint to update folders
@router.put(
    "/folders",
    summary="Update the folder lists",
    status_code=status.HTTP_202_ACCEPTED,
)
def enqueue_update_folders(
    folders: Folders = Body(
        ...,
        title="The new sets of included and excluded folders. Replaces the current lists with these.",
    ),
    conn_args: Dict[str, Any] = Depends(get_db_system_wl),
) -> JobModel:
    queue_id = job_manager.get_next_job_id()
    job = Job(
        queue_id=queue_id,
        job_type="folder_update",
        conn_args=conn_args,
        included_folders=folders.included_folders,
        excluded_folders=folders.excluded_folders,
    )
    job_manager.enqueue_job(job)
    return JobModel(queue_id=job.queue_id, job_type=job.job_type, metadata=None)


class QueueCancelResponse(BaseModel):
    cancelled_jobs: List[int]


# Endpoint to cancel queued jobs
@router.delete(
    "/queue",
    summary="Cancel queued jobs",
    status_code=status.HTTP_200_OK,
)
def cancel_queued_jobs(
    queue_ids: List[int] = Query(..., title="List of Queue IDs to cancel"),
) -> QueueCancelResponse:
    cancelled = job_manager.cancel_queued_jobs(queue_ids)
    if not cancelled:
        raise HTTPException(
            status_code=404, detail="No matching queued jobs found."
        )
    return QueueCancelResponse(cancelled_jobs=cancelled)


class CancelResponse(BaseModel):
    detail: str


# Endpoint to cancel the currently running job
@router.post(
    "/cancel",
    summary="Cancel the currently running job",
    status_code=status.HTTP_200_OK,
)
def cancel_current_job() -> CancelResponse:
    cancelled_job_id = job_manager.cancel_running_job()
    if cancelled_job_id is None:
        raise HTTPException(
            status_code=404, detail="No job is currently running."
        )
    return CancelResponse(detail=f"Job {cancelled_job_id} cancelled.")


# Additional endpoints remain unchanged
@router.get(
    "/folders",
    summary="Get the current folder lists",
)
def get_folders(
    conn_args: Dict[str, Any] = Depends(get_db_readonly),
) -> Folders:
    conn = get_database_connection(**conn_args)
    try:
        current_included_folders = get_folders_from_database(
            conn, included=True
        )
        current_excluded_folders = get_folders_from_database(
            conn, included=False
        )
        return Folders(
            included_folders=current_included_folders,
            excluded_folders=current_excluded_folders,
        )
    finally:
        conn.close()


@router.get(
    "/folders/history",
    summary="Get the scan history",
)
def get_scan_history(
    conn_args: Dict[str, Any] = Depends(get_db_readonly),
) -> List[FileScanRecord]:
    conn = get_database_connection(**conn_args)
    try:
        return get_all_file_scans(conn)
    finally:
        conn.close()


@router.get(
    "/data/history",
    summary="Get the extraction history",
)
def get_extraction_history(
    conn_args: Dict[str, Any] = Depends(get_db_readonly),
) -> List[LogRecord]:
    conn = get_database_connection(**conn_args)
    try:
        return get_all_data_logs(conn)
    finally:
        conn.close()


class ConfigResponse(BaseModel):
    detail: str


@router.put(
    "/config",
    summary="Update the system configuration",
    status_code=status.HTTP_200_OK,
)
def update_config(
    config: SystemConfig = Body(..., title="The new system configuration"),
    conn_args: Dict[str, Any] = Depends(get_db_system_wl),
) -> ConfigResponse:
    conn = get_database_connection(**conn_args)
    try:
        persist_system_config(conn, config)
        conn.commit()
    finally:
        conn.close()
    return ConfigResponse(detail="System configuration updated.")


@router.get(
    "/config",
    summary="Get the current system configuration",
    response_model=SystemConfig,
)
def get_config(
    conn_args: Dict[str, Any] = Depends(get_db_readonly),
) -> SystemConfig:
    conn = get_database_connection(**conn_args)
    try:
        return retrieve_system_config(conn)
    finally:
        conn.close()


# To support forward references in Pydantic models
QueueStatusModel.model_rebuild()
RunningJobModel.model_rebuild()
JobModel.model_rebuild()

File: gateway/src/api/jobs.rs
use axum::{Json, http::StatusCode};
// axum's own Query (serde_urlencoded) cannot deserialize repeated params
// (?inference_ids=a&inference_ids=b) into a Vec; axum-extra's can, matching
// FastAPI's List[str] query parameter behavior.
use axum_extra::extract::Query;
use serde::Deserialize;
use utoipa::{IntoParams, ToSchema};

use crate::api::db_params::DbQueryParams;
use crate::api_error::ApiError;
use crate::db::extraction_log::{LogRecord, get_all_data_logs, get_setters_total_data};
use crate::db::file_scans::get_all_file_scans;
use crate::db::folders::get_folders_from_database;
use crate::db::system_config::{SystemConfig, SystemConfigStore};
use crate::db::{DbConnection, ReadOnly};
use crate::jobs::continuous_scan;
use crate::jobs::cron::{self, CronRunOutcome};
use crate::jobs::files::is_resync_needed;
use crate::jobs::queue::{
    JobModel, JobRequest, JobType, QueueStatusModel, cancel_queued_jobs, cancel_running_job,
    enqueue_job, get_queue_status,
};

#[derive(Deserialize, IntoParams)]
#[into_params(parameter_in = Query)]
pub(crate) struct InferenceQuery {
    /// Inference ID List
    inference_ids: Vec<String>,
    /// Batch Size
    batch_size: Option<i64>,
    /// Confidence Threshold
    threshold: Option<f64>,
}

#[derive(Deserialize, IntoParams)]
#[into_params(parameter_in = Query)]
pub(crate) struct LogIdQuery {
    /// List of Log Ids to delete the generated data for
    log_ids: Vec<i64>,
}

#[derive(Deserialize, IntoParams)]
#[into_params(parameter_in = Query)]
pub(crate) struct QueueCancelQuery {
    /// List of Queue IDs to cancel
    queue_ids: Vec<i64>,
}

#[derive(Deserialize, IntoParams)]
#[into_params(parameter_in = Query)]
pub(crate) struct HistoryQuery {
    /// Page number
    #[param(default = 1, minimum = 1)]
    page: Option<i64>,
    /// Page size
    #[param(minimum = 1)]
    page_size: Option<i64>,
}

#[derive(serde::Serialize, ToSchema)]
pub(crate) struct QueueCancelResponse {
    cancelled_jobs: Vec<i64>,
}

#[derive(serde::Serialize, ToSchema)]
pub(crate) struct CancelResponse {
    detail: String,
}

#[derive(serde::Serialize, ToSchema)]
pub(crate) struct FoldersResponse {
    included_folders: Vec<String>,
    excluded_folders: Vec<String>,
}

#[derive(serde::Serialize, ToSchema)]
pub(crate) struct SetterDataStats {
    total_counts: Vec<(String, i64)>,
}

#[derive(serde::Serialize, ToSchema)]
pub(crate) struct CronJobResponse {
    detail: String,
}

#[utoipa::path(
    get,
    path = "/api/jobs/queue",
    tag = "jobs",
    summary = "Get running job and queue status",
    responses(
        (status = 200, description = "Queue status", body = QueueStatusModel)
    )
)]
pub(crate) async fn queue_status() -> Result<Json<QueueStatusModel>, ApiError> {
    let status = get_queue_status().await?;
    Ok(Json(status))
}

#[utoipa::path(
    post,
    path = "/api/jobs/data/extraction",
    tag = "jobs",
    summary = "Run a data extraction job",
    params(DbQueryParams, InferenceQuery),
    responses(
        (status = 202, description = "Enqueued data extraction jobs", body = [JobModel])
    )
)]
pub(crate) async fn enqueue_data_extraction(
    Query(query): Query<InferenceQuery>,
    conn: DbConnection<ReadOnly>,
) -> Result<(StatusCode, Json<Vec<JobModel>>), ApiError> {
    // Validate the models and resolve effective batch_size/threshold at
    // enqueue time (mirrors Python): a bad inference ID fails this request
    // instead of a job hours later, and the queue status shows the values
    // the job will actually run with.
    let store = SystemConfigStore::from_env();
    let config = store.load(&conn.index_db)?;
    let mut jobs = Vec::new();
    for inference_id in query.inference_ids {
        let model = crate::jobs::extraction::load_model_metadata(&inference_id).await?;
        let defaults = crate::jobs::extraction::resolve_job_defaults(
            &config,
            &model,
            query.batch_size,
            query.threshold,
        );
        let job = enqueue_job(JobRequest {
            job_type: JobType::DataExtraction,
            index_db: conn.index_db.clone(),
            user_data_db: conn.user_data_db.clone(),
            metadata: Some(inference_id),
            batch_size: Some(defaults.batch_size),
            threshold: defaults.threshold,
            log_id: None,
            tag: None,
        })
        .await?;
        jobs.push(job);
    }
    Ok((StatusCode::ACCEPTED, Json(jobs)))
}

#[utoipa::path(
    delete,
    path = "/api/jobs/data/extraction",
    tag = "jobs",
    summary = "Delete extracted data",
    params(DbQueryParams, InferenceQuery),
    responses(
        (status = 202, description = "Enqueued data deletion jobs", body = [JobModel])
    )
)]
pub(crate) async fn enqueue_delete_extracted_data(
    Query(query): Query<InferenceQuery>,
    conn: DbConnection<ReadOnly>,
) -> Result<(StatusCode, Json<Vec<JobModel>>), ApiError> {
    let mut jobs = Vec::new();
    for inference_id in query.inference_ids {
        let job = enqueue_job(JobRequest {
            job_type: JobType::DataDeletion,
            index_db: conn.index_db.clone(),
            user_data_db: conn.user_data_db.clone(),
            metadata: Some(inference_id),
            batch_size: None,
            threshold: None,
            log_id: None,
            tag: None,
        })
        .await?;
        jobs.push(job);
    }
    Ok((StatusCode::ACCEPTED, Json(jobs)))
}

#[utoipa::path(
    post,
    path = "/api/jobs/folders/rescan",
    tag = "jobs",
    summary = "Run a folder rescan",
    params(DbQueryParams),
    responses(
        (status = 202, description = "Enqueued folder rescan job", body = JobModel)
    )
)]
pub(crate) async fn enqueue_folder_rescan(
    conn: DbConnection<ReadOnly>,
) -> Result<(StatusCode, Json<JobModel>), ApiError> {
    let job = enqueue_job(JobRequest {
        job_type: JobType::FolderRescan,
        index_db: conn.index_db.clone(),
        user_data_db: conn.user_data_db.clone(),
        metadata: None,
        batch_size: None,
        threshold: None,
        log_id: None,
        tag: None,
    })
    .await?;
    Ok((StatusCode::ACCEPTED, Json(job)))
}

#[utoipa::path(
    put,
    path = "/api/jobs/folders",
    tag = "jobs",
    summary = "Update the database with the current folder lists in the config",
    description = "Must be run every time after the folder lists in the config are updated,\nto ensure that the database is in sync with the config.\nIf you update the config through the API, this will be done automatically if needed.\n\nThis will remove files and items from the database that are no longer in the included folders,\nand add files and items that are now in the included folders, as well as remove files and items\nfrom the database that are now in the excluded folders.",
    params(DbQueryParams),
    responses(
        (status = 202, description = "Enqueued folder update job", body = JobModel)
    )
)]
pub(crate) async fn enqueue_update_folders(
    conn: DbConnection<ReadOnly>,
) -> Result<(StatusCode, Json<JobModel>), ApiError> {
    let job = enqueue_job(JobRequest {
        job_type: JobType::FolderUpdate,
        index_db: conn.index_db.clone(),
        user_data_db: conn.user_data_db.clone(),
        metadata: None,
        batch_size: None,
        threshold: None,
        log_id: None,
        tag: None,
    })
    .await?;
    Ok((StatusCode::ACCEPTED, Json(job)))
}

#[utoipa::path(
    delete,
    path = "/api/jobs/queue",
    tag = "jobs",
    summary = "Cancel queued jobs",
    params(QueueCancelQuery),
    responses(
        (status = 200, description = "Queued jobs cancelled", body = QueueCancelResponse)
    )
)]
pub(crate) async fn cancel_queued(
    Query(query): Query<QueueCancelQuery>,
) -> Result<Json<QueueCancelResponse>, ApiError> {
    let cancelled = cancel_queued_jobs(query.queue_ids).await?;
    if cancelled.is_empty() {
        return Err(ApiError::not_found("No matching queued jobs found."));
    }
    Ok(Json(QueueCancelResponse {
        cancelled_jobs: cancelled,
    }))
}

#[utoipa::path(
    post,
    path = "/api/jobs/cancel",
    tag = "jobs",
    summary = "Cancel the currently running job",
    responses(
        (status = 200, description = "Running job cancelled", body = CancelResponse)
    )
)]
pub(crate) async fn cancel_current_job() -> Result<Json<CancelResponse>, ApiError> {
    let cancelled = cancel_running_job().await?;
    let job_id = cancelled.ok_or_else(|| ApiError::not_found("No job is currently running."))?;
    Ok(Json(CancelResponse {
        detail: format!("Job {job_id} cancelled."),
    }))
}

#[utoipa::path(
    get,
    path = "/api/jobs/folders",
    tag = "jobs",
    summary = "Get the current folder lists",
    description = "Get the current included and excluded folders in the database.\nThese are the folders that are being scanned and not being scanned, respectively.\n\nThis list may differ from the config, if the database has not been updated.",
    params(DbQueryParams),
    responses(
        (status = 200, description = "Current folder lists", body = FoldersResponse)
    )
)]
pub(crate) async fn get_folders(
    mut conn: DbConnection<ReadOnly>,
) -> Result<Json<FoldersResponse>, ApiError> {
    let included = get_folders_from_database(&mut conn.conn, true).await?;
    let excluded = get_folders_from_database(&mut conn.conn, false).await?;
    Ok(Json(FoldersResponse {
        included_folders: included,
        excluded_folders: excluded,
    }))
}

#[utoipa::path(
    get,
    path = "/api/jobs/folders/history",
    tag = "jobs",
    summary = "Get the scan history",
    params(DbQueryParams, HistoryQuery),
    responses(
        (status = 200, description = "Scan history", body = [crate::db::file_scans::FileScanRecord])
    )
)]
pub(crate) async fn get_scan_history(
    Query(query): Query<HistoryQuery>,
    mut conn: DbConnection<ReadOnly>,
) -> Result<Json<Vec<crate::db::file_scans::FileScanRecord>>, ApiError> {
    let page = query.page.unwrap_or(1);
    let scans = get_all_file_scans(&mut conn.conn, page, query.page_size).await?;
    Ok(Json(scans))
}

#[utoipa::path(
    delete,
    path = "/api/jobs/data/history",
    tag = "jobs",
    summary = "Deletes data generated by the scans given log ids",
    params(DbQueryParams, LogIdQuery),
    responses(
        (status = 200, description = "Enqueued data deletion jobs", body = [JobModel])
    )
)]
pub(crate) async fn delete_scan_data(
    Query(query): Query<LogIdQuery>,
    conn: DbConnection<ReadOnly>,
) -> Result<Json<Vec<JobModel>>, ApiError> {
    let mut jobs = Vec::new();
    for log_id in query.log_ids {
        let job = enqueue_job(JobRequest {
            job_type: JobType::JobDataDeletion,
            index_db: conn.index_db.clone(),
            user_data_db: conn.user_data_db.clone(),
            metadata: None,
            batch_size: None,
            threshold: None,
            log_id: Some(log_id),
            tag: None,
        })
        .await?;
        jobs.push(job);
    }
    Ok(Json(jobs))
}

#[utoipa::path(
    get,
    path = "/api/jobs/data/history",
    tag = "jobs",
    summary = "Get the extraction history",
    params(DbQueryParams, HistoryQuery),
    responses(
        (status = 200, description = "Extraction history", body = [LogRecord])
    )
)]
pub(crate) async fn get_extraction_history(
    Query(query): Query<HistoryQuery>,
    mut conn: DbConnection<ReadOnly>,
) -> Result<Json<Vec<LogRecord>>, ApiError> {
    let page = query.page.unwrap_or(1);
    let logs = get_all_data_logs(&mut conn.conn, page, query.page_size).await?;
    Ok(Json(logs))
}

#[utoipa::path(
    put,
    path = "/api/jobs/config",
    tag = "jobs",
    summary = "Update the system configuration",
    params(DbQueryParams),
    request_body(content = SystemConfig, description = "The new system configuration"),
    responses(
        (status = 200, description = "Updated system configuration", body = SystemConfig)
    )
)]
pub(crate) async fn update_config(
    conn: DbConnection<ReadOnly>,
    Json(config): Json<SystemConfig>,
) -> Result<Json<SystemConfig>, ApiError> {
    // Python accepts unparseable cron strings and fails invisibly inside the
    // scheduler forever; reject them here so typos surface at save time.
    if let Err(err) = cron::validate_cron_schedule(&config.cron_schedule) {
        return Err(ApiError::bad_request(format!(
            "Invalid cron_schedule {:?}: {err}",
            config.cron_schedule
        )));
    }
    let store = SystemConfigStore::from_env();
    store.save(&conn.index_db, &config)?;
    let config = store.load(&conn.index_db)?;
    let _ = continuous_scan::notify_config_change(&conn.index_db).await;
    let _ = cron::notify_config_change(&conn.index_db).await;
    let resync_needed = is_resync_needed(&conn.index_db, &conn.user_data_db, &config).await?;
    if resync_needed {
        let _ = enqueue_job(JobRequest {
            job_type: JobType::FolderUpdate,
            index_db: conn.index_db.clone(),
            user_data_db: conn.user_data_db.clone(),
            metadata: None,
            batch_size: None,
            threshold: None,
            log_id: None,
            tag: None,
        })
        .await?;
    }
    Ok(Json(config))
}

#[utoipa::path(
    get,
    path = "/api/jobs/config",
    tag = "jobs",
    summary = "Get the current system configuration",
    params(DbQueryParams),
    responses(
        (status = 200, description = "Current system configuration", body = SystemConfig)
    )
)]
pub(crate) async fn get_config(
    conn: DbConnection<ReadOnly>,
) -> Result<Json<SystemConfig>, ApiError> {
    let store = SystemConfigStore::from_env();
    let config = store.load(&conn.index_db)?;
    Ok(Json(config))
}

#[utoipa::path(
    get,
    path = "/api/jobs/data/setters/total",
    tag = "jobs",
    summary = "Get the total count of index data entry for each setter",
    params(DbQueryParams),
    responses(
        (status = 200, description = "Total setter data counts", body = SetterDataStats)
    )
)]
pub(crate) async fn get_setter_data_count(
    mut conn: DbConnection<ReadOnly>,
) -> Result<Json<SetterDataStats>, ApiError> {
    let totals = get_setters_total_data(&mut conn.conn).await?;
    Ok(Json(SetterDataStats {
        total_counts: totals,
    }))
}

#[utoipa::path(
    post,
    path = "/api/jobs/cronjob/run",
    tag = "jobs",
    summary = "Manually trigger a cronjob run",
    description = "Manually trigger the configured cronjob to run on the selected database.",
    params(DbQueryParams),
    responses(
        (status = 200, description = "Cronjob triggered", body = CronJobResponse)
    )
)]
pub(crate) async fn manual_trigger_cronjob(
    conn: DbConnection<ReadOnly>,
) -> Result<Json<CronJobResponse>, ApiError> {
    let detail = match cron::run_cronjob(&conn.index_db, &conn.user_data_db).await? {
        CronRunOutcome::Enqueued(_) => "Cronjob triggered.".to_string(),
        // Python also replies 200 here (the skip is silent); keep the status
        // code but say what happened.
        CronRunOutcome::Skipped => {
            "Cronjob skipped: a previous cronjob for this database is still queued or running."
                .to_string()
        }
    };
    Ok(Json(CronJobResponse { detail }))
}

#[derive(serde::Serialize, ToSchema)]
pub(crate) struct CronScheduleResponse {
    /// Whether automatic cron runs are enabled for this database.
    enabled: bool,
    /// The configured cron schedule string.
    cron_schedule: String,
    /// Whether the configured schedule string parses.
    valid: bool,
    /// Next automatic run (RFC 3339, local time), when scheduling is active.
    next_run: Option<String>,
    /// Last automatic run fired by this process (RFC 3339, local time).
    /// Manual triggers are not included.
    last_run: Option<String>,
}

#[utoipa::path(
    get,
    path = "/api/jobs/cronjob/schedule",
    tag = "jobs",
    summary = "Get the cronjob schedule status",
    description = "Get the configured cron schedule for the selected database along with the next and last automatic run times.",
    params(DbQueryParams),
    responses(
        (status = 200, description = "Cronjob schedule status", body = CronScheduleResponse)
    )
)]
pub(crate) async fn get_cronjob_schedule(
    conn: DbConnection<ReadOnly>,
) -> Result<Json<CronScheduleResponse>, ApiError> {
    let store = SystemConfigStore::from_env();
    let config = store.load(&conn.index_db)?;
    let status = cron::get_schedule_status(&conn.index_db)
        .await
        .unwrap_or_default();
    Ok(Json(CronScheduleResponse {
        enabled: config.enable_cron_job,
        valid: cron::validate_cron_schedule(&config.cron_schedule).is_ok(),
        cron_schedule: config.cron_schedule,
        next_run: status.next_run.map(|time| time.to_rfc3339()),
        last_run: status.last_run.map(|time| time.to_rfc3339()),
    }))
}

#[cfg(test)]
mod tests {
    use super::*;
    use axum::http::Uri;

    /// The UI sends list params FastAPI-style
    /// (?inference_ids=a&inference_ids=b). Plain axum::extract::Query
    /// (serde_urlencoded) rejects repeated keys into a Vec with a 400, so
    /// these structs must go through axum-extra's Query.
    #[test]
    fn repeated_query_params_parse_into_vecs() {
        let uri: Uri = "/api/jobs/data/extraction\
            ?inference_ids=tags/wd-swinv2-tagger-v3\
            &inference_ids=clip/ViT-H-14-378-quickgelu_dfn5b\
            &batch_size=64"
            .parse()
            .unwrap();
        let Query(q) = Query::<InferenceQuery>::try_from_uri(&uri).unwrap();
        assert_eq!(
            q.inference_ids,
            vec!["tags/wd-swinv2-tagger-v3", "clip/ViT-H-14-378-quickgelu_dfn5b"]
        );
        assert_eq!(q.batch_size, Some(64));
        assert_eq!(q.threshold, None);

        let uri: Uri = "/x?log_ids=1&log_ids=2".parse().unwrap();
        let Query(q) = Query::<LogIdQuery>::try_from_uri(&uri).unwrap();
        assert_eq!(q.log_ids, vec![1, 2]);

        let uri: Uri = "/x?queue_ids=3".parse().unwrap();
        let Query(q) = Query::<QueueCancelQuery>::try_from_uri(&uri).unwrap();
        assert_eq!(q.queue_ids, vec![3]);
    }
}
